File: sparksqltest20/src/main/scala/com/qi/UserStastictics.java
package com.qi;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

public class UserStastictics {

    //对每个用户,统计login次数,logout次数,view_user次数,new_tweet次数
    //每个ip上平均行为次数=用户行为次数/该用户使用过的ip数量

    public static SparkConf conf = new SparkConf().setAppName("UserStatic").setMaster("local[*]");

    public static JavaSparkContext sc = new JavaSparkContext(conf);

    public static void loginNumber() {
        JavaRDD<String> rdd = sc.textFile("file:///C:\\Users\\Administrator\\Desktop\\user-logs-large.txt");

        //1
        JavaPairRDD<String, String> rdd1 = rdd.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String s) throws Exception {
                String[] strings = s.split("\\s+");
                String s1 = strings[0] + "\t" + strings[1];
                return Arrays.asList(s1).iterator();
            }
        }).mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        }).mapToPair(new PairFunction<Tuple2<String, Integer>, String, String>() {
            @Override
            public Tuple2<String, String> call(Tuple2<String, Integer> stringIntegerTuple2) throws Exception {
                String[] split = stringIntegerTuple2._1.split("\\s+");

                return new Tuple2<String, String>(split[0], split[1] + "->" + stringIntegerTuple2._2);
            }
        }).reduceByKey(new Function2<String, String, String>() {
            @Override
            public String call(String s, String s2) throws Exception {
                return s + "\t" + s2;
            }
        });

        //2
        JavaPairRDD<String, Integer> user = rdd.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String s) throws Exception {
                String[] strings = s.split("\\s+");
                return Arrays.asList(strings[0]).iterator();
            }
        }).mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer integer, Integer integer2) throws Exception {
                return integer + integer2;
            }
        });

        //3
        JavaPairRDD<String, Integer> userIps = rdd.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String s) throws Exception {
                String[] split = s.split("\\s+");
                String s1 = split[0] + "\t" + split[2];
                return Arrays.asList(s1).iterator();
            }
        }).distinct().mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {

                return new Tuple2<String, Integer>(s.split("\\s+")[0], 1);
            }
        }).reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer integer, Integer integer2) throws Exception {
                return integer + integer2;
            }
        });

        //4
        JavaPairRDD<String, Integer> ips = user.join(userIps).mapToPair(new PairFunction<Tuple2<String, Tuple2<Integer, Integer>>, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(Tuple2<String, Tuple2<Integer, Integer>> stringTuple2Tuple2) throws Exception {
                return new Tuple2<String, Integer>(stringTuple2Tuple2._1, stringTuple2Tuple2._2._1 / stringTuple2Tuple2._2._2);
            }
        });

        //5结果
        rdd1.join(ips).mapToPair(new PairFunction<Tuple2<String, Tuple2<String, Integer>>, String, String>() {
            @Override
            public Tuple2<String, String> call(Tuple2<String, Tuple2<String, Integer>> s2) throws Exception {

                return new Tuple2<String, String>(s2._1, s2._2._1 + "\t" + "ip" + "->" + s2._2._2);
            }
        }).foreach(new VoidFunction<Tuple2<String, String>>() {
            @Override
            public void call(Tuple2<String, String> stringStringTuple2) throws Exception {
                System.out.println(stringStringTuple2._1 + "\t" + stringStringTuple2._2);
            }
        });
    }

    public static void main(String[] args) {
        loginNumber();
    }
}

File: sparkTest/src/main/scala/com/qi/spark_test/rdd_test_teacher/UserStastictics.scala
package com.qi.spark_test.rdd_test_teacher

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created by ThinkPad on 2018/4/4.
  */
case class UserLogAction(name:String, actionType:String, ip:String)
object UserStastictics {
  val conf = new SparkConf().setMaster("local[*]").setAppName("useraction")
  val sc = SparkContext.getOrCreate(conf)

  def userActionTimes(inputPath:String) = {
    //加载输入
    val rdd = sc.textFile(inputPath, 3)
    val parserRdd =   rdd.map(x=>{
      val reg = "(.*)\t(.*)\t(.*)".r
      val actionLog = x match{
        case reg(name,actionType,ip) => UserLogAction(name,actionType,ip)
        case _ => null
      }
      actionLog
    }).map(x=>(x.name,x))
//    parserRdd.take(10).foreach(println)
    parserRdd.cache()
    //统计login次数，logout次数，view_user次数，new_tweet次数
    val times = parserRdd.combineByKey(
      x=>(if(x.actionType=="login") 1 else 0,
        if(x.actionType=="logout") 1 else 0,
        if(x.actionType=="view_user") 1 else 0,
        if(x.actionType=="new_tweet") 1 else 0)

      ,(c:(Int,Int,Int,Int),v:UserLogAction)=>{
        (c._1+(if(v.actionType=="login") 1 else 0),
          c._2+(if(v.actionType=="logout") 1 else 0),
          c._3+(if(v.actionType=="view_user") 1 else 0),
          c._4+(if(v.actionType=="new_tweet") 1 else 0)
        )
      }

      ,(c1:(Int,Int,Int,Int),c2:(Int,Int,Int,Int))=>(c1._1+c2._1,
        c1._2+c2._2,
        c1._3+c2._3,
        c1._4+c2._4)
    )
//    times.foreach(println)
    //每个ip上平均行为次数= 用户行为次数/该用户使用过的ip数量
    //统计用户的行为次数
    val actionTimes = parserRdd.map(x=>(x._1,1)).reduceByKey(_+_)
    val userIps = parserRdd.map(x=>(x._1,x._2.ip)).distinct().map(x=>(x._1,1)).reduceByKey(_+_)
    val userIpAvgTimes = actionTimes.join(userIps).map(x=>(x._1,x._2._1*1.0/x._2._2))
    val result = times.join(userIpAvgTimes)

    result.map(x=>(x._1,x._2._1._1,x._2._1._2,x._2._1._3,x._2._1._4,x._2._2))
    .saveAsTextFile("file:///e:/useractiontimes")
  }

  def main(args: Array[String]): Unit = {
    userActionTimes("file:///E:\\spartest\\user-logs-large.txt")
  }
}
